### PR TITLE
Copy location

### PR DESCRIPTION
**Pull request by CalebKoch**
Intended to address #697 

Still have to make sure it copies over Aura's, trackers, labels, and groups. But wanted to send this out a little early to see if I could get any feedback. I feel like there should be a better way to do the make_copy methods, but I'm not super familiar with peewee, and I couldn't find anything in the documentation that would seem to help.

### CODE REVIEW

**Comment by Kruptein** [comment]
Hey thanks for picking this up, I first need to do some hotfix work, I'll see if I can manage to get a review in today, otherwise I'll tackle this wednesday :)

**Comment by CalebKoch** [comment]
Sounds good. 
I just got Aura's and trackers added. My initial search didn't return where I need to would update for Groups, and Label's seems structured a little different. I'm pretty busy the rest of the week, so I probably won't be able to dig into those last two issues untill next Monday, but I would still appreciate an early review.

**Comment by CalebKoch** [comment]
Just added copying labels, and fixing Groups. Should be ready to go, whenever you have some time to review.

**Comment by CalebKoch** [comment]
This pr should be good to go whenever you have a chance!

File: client/src/game/store.ts [hunk 1]
```diff
@@ -341,6 +342,15 @@ class GameStore extends VuexModule implements GameState {
         if (data.sync) sendLocationUnarchive(data.id);
     }
 
+    @Mutation
+    cloneLocation(data: { location: number; room: string; sync: boolean }): void {
+        const location = this.locations.find((l) => l.id === data.location);
+        if (location === undefined) {
+            throw new Error("unknown location rename attempt");
```
Comment (Kruptein) [inline client/src/game/store.ts:349]:
```suggestion
            throw new Error("unknown location clone attempt");
```

File: client/src/game/ui/settings/location/LocationAdminSettings.vue [hunk 2]
```diff
@@ -53,12 +61,40 @@ export default class LocationAdminSettings extends Vue {
...
+
+            gameStore.cloneLocation({
+                location: this.location,
+                room: roomName,
+                sync: true,
+            });
+
+            this.$emit("close");
+        } else {
+            // TODO: Handle error codes. Don't see any type of Notification modal. Should it just console.log?
```
Comment (Kruptein) [inline client/src/game/ui/settings/location/LocationAdminSettings.vue:88]: Keep it as a console log for now

File: server/api/socket/location.py [hunk 3]
```diff
@@ -358,6 +367,36 @@ async def add_new_location(sid: str, location: str):
...
+        logger.warning(f"{pr.player.name} attempted to clone locations.")
+        return
+    try:
+        room = (
+            Room.select().where(
+                (Room.name == data["room"]) & (Room.creator == pr.player)
+            )[0]
+        )
+    except IndexError:
+        logger.warning(f"Destination room %s not found.", data["room"])
```
Comment (Kruptein) [inline server/api/socket/location.py:384]:
```suggestion
        logger.warning(f"Destination room {data['room']} not found.")
```

File: server/models/shape/__init__.py [hunk 4]
```diff
@@ -118,6 +119,52 @@ def center_at(self, x: int, y: int) -> None:
...
+            annotation=self.annotation,
+            draw_operator=self.draw_operator,
+            index=self.index,
+            options=self.options,
+            badge=self.badge,
+            show_badge=self.show_badge,
+            default_edit_access=self.default_edit_access,
+            default_vision_access=self.default_vision_access,
+            is_invisible=self.is_invisible,
+            default_movement_access=self.default_movement_access,
```
Comment (Kruptein) [inline server/models/shape/__init__.py:145]: is_defeated is missing here

File: server/models/shape/__init__.py [hunk 5]
```diff
@@ -118,6 +119,52 @@ def center_at(self, x: int, y: int) -> None:
...
+            is_locked=self.is_locked,
+            angle=self.angle,
+            stroke_width=self.stroke_width,
+            asset=self.asset,
+            group=None if self.group is None else self.group.make_copy(),
```
Comment (Kruptein) [inline server/models/shape/__init__.py:150]: This creates a new group for each member. So for example a group with 3 shapes will result in 3 separate groups

File: server/models/shape/__init__.py [hunk 6]
```diff
@@ -144,6 +196,11 @@ def __repr__(self):
     def as_dict(self):
         return model_to_dict(self, recurse=False, exclude=[Tracker.shape])
 
+    def make_copy(self, new_shape):
+        dict = self.as_dict()
+        dict['uuid'] = str(uuid4())
+        copy = type(self).create(shape=new_shape, **dict)
```
Comment (Kruptein) [inline server/models/shape/__init__.py:202]:
```suggestion
        type(self).create(shape=new_shape, **dict)
```

File: server/models/shape/__init__.py [hunk 7]
```diff
@@ -165,6 +222,11 @@ def __repr__(self):
     def as_dict(self):
         return model_to_dict(self, recurse=False, exclude=[Aura.shape])
 
+    def make_copy(self, new_shape):
+        dict = self.as_dict()
+        dict['uuid'] = str(uuid4())
+        copy = type(self).create(shape=new_shape, **dict)
```
Comment (Kruptein) [inline server/models/shape/__init__.py:228]:
```suggestion
        type(self).create(shape=new_shape, **dict)
```

File: server/models/shape/__init__.py [hunk 8]
```diff
@@ -212,6 +274,11 @@ def get_center_offset(self, x: int, y: int) -> Tuple[int, int]:
     def set_location(self, points: List[List[int]]) -> None:
         logger.error("Attempt to set location on shape without location info")
 
+    def make_copy(self, new_shape):
+        table = type(self)
+        dict = self.as_dict(exclude=[table.shape])
+        copy = table.create(shape=new_shape, **dict)
```
Comment (Kruptein) [inline server/models/shape/__init__.py:280]:
```suggestion
        table.create(shape=new_shape, **dict)
```

**Review by Kruptein — CHANGES_REQUESTED** [review]
Good job, some minor things and a bug with groups, but otherwise works great!

There is one special shape that will not properly be copied currently (ToggleComposite) because it has uuid references to other shapes that are not tracked, but I'm fine with leaving that out of scope

File: server/models/shape/__init__.py (hunk 5 above)
Comment (Kruptein) [reply server/models/shape/__init__.py:151]: This hasn't been addressed yet

Comment (CalebKoch) [reply server/models/shape/__init__.py]: Yeah, had a couple busy weekends, and it took a bit to come up with a good way to make this work. Just pushed something up that takes care of it.

**Pull request merged**